subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c

```diff
@@ -0,0 +1,1281 @@
+	sys_bitarray_clear_bit(conf->rx_hold_bitmap, rx_block_index);
+
+	/* Call the endpoint callback. It can set the hold bit. */
+	ept->cfg->cb.received(buffer, size, ept->cfg->priv);
+
+	/* If the bit is still cleared, request release of the buffer. */
+	sys_bitarray_test_bit(conf->rx_hold_bitmap, rx_block_index, &bit_val);
+	if (!bit_val) {
+		send_release(dev_data, buffer, MSG_RELEASE_DATA, 0);
+	}
+
+	return 0;
+}
+
+/**
+ * Release data message received.
+ */
+static int received_release_data(struct backend_data *dev_data, size_t tx_block_index,
+				 uint8_t ept_addr)
+{
+	const struct icmsg_with_buf_config *conf = dev_data->conf;
+	uint8_t *buffer;
+	struct ept_data *ept;
+	size_t size;
+	int r;
+
+	/* Validate. */
+	buffer = buffer_from_index_validate(&conf->tx, tx_block_index, &size, false);
+	if (buffer == NULL) {
+		LOG_ERR("Received invalid block index %d", tx_block_index);
+		return -EINVAL;
+	}
+
+	/* Release. */
+	r = release_tx_blocks(dev_data, tx_block_index, size, -1);
+	if (r < 0) {
+		return r;
+	}
+
+	/* When releasing bound message, validate RX state of the endpoint. */
+	if (ept_addr != EPT_ADDR_INVALID) {
+		ept = get_ept_and_rx_validate(dev_data, ept_addr);
+		if (ept == NULL) {
+			LOG_ERR("Received invalid endpoint address %d", ept_addr);
+			r = -EINVAL;
+		}
+	}
+
+	return r;
+}
+
+/**
+ * Bound endpoint message received.
+ */
+static int received_bound(struct backend_data *dev_data, size_t rx_block_index,
+			  uint8_t ept_addr)
+{
+	const struct icmsg_with_buf_config *conf = dev_data->conf;
+	size_t size;
+	uint8_t *buffer;
+
+	/* Validate */
+	buffer = buffer_from_index_validate(&conf->rx, rx_block_index, &size, true);
+	if (buffer == NULL) {
+		LOG_ERR("Received invalid block index %d", rx_block_index);
+		return -EINVAL;
+	}
+
+	/* Put message to waiting array. */
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+	dev_data->waiting_bound[ept_addr] = rx_block_index;
+	k_mutex_unlock(&dev_data->mutex);
+
+	/* Schedule processing the message. */
+	schedule_ept_bound_process(dev_data);
+
+	return 0;
+}
+
+/**
+ * Callback called by ICMsg that handles message (data or endpoint bound) received
+ * from the remote.
+ *
+ * @param[in] data	Message received from the ICMsg.
+ * @param[in] len	Number of bytes of data.
+ * @param[in] priv	Opaque pointer to device instance.
+ */
+static void received(const void *data, size_t len, void *priv)
+{
+	const struct device *instance = priv;
+	struct backend_data *dev_data = instance->data;
+	const struct icmsg_message *message = (const struct icmsg_message *)data;
+	uint8_t ept_addr;
+	int r = 0;
+
+	/* Allow messages longer than 3 bytes, e.g. for future protocol versions. */
+	if (len < sizeof(struct icmsg_message)) {
+		r = -EINVAL;
+		goto exit;
+	}
+
+	ept_addr = message->ept_addr;
+	if (ept_addr >= NUM_EPT) {
+		r = -EINVAL;
+		goto exit;
+	}
+
+	switch (message->msg_type) {
+	case MSG_RELEASE_DATA:
+		ept_addr = EPT_ADDR_INVALID;
+		__fallthrough;
+	case MSG_RELEASE_BOUND:
```

Comment:
Is there any difference between `MSG_RELEASE_BOUND` and `MSG_RELEASE_DATA`? Could we simplify it just to `MSG_RELEASE`

Reply:
https://github.com/zephyrproject-rtos/zephyr/pull/58741#discussion_r1357951890